src/vm/precode.cpp

```diff
@@ -836,75 +796,103 @@ static void SaveFixupPrecodeChunk(DataImage * image, MethodDesc ** rgMD, COUNT_T
 {
     STANDARD_VM_CONTRACT;
 
-#ifdef HAS_FIXUP_PRECODE_CHUNKS
-    if (m_rgPendingChunk.GetCount() == 0)
-        return;
+    ULONG sizeOfOne = Precode::SizeOfTemporaryEntryPoint(PRECODE_STUB);
+    ULONG size = sizeOfOne * count;
+    TADDR pBase = (TADDR)new (image->GetHeap()) BYTE[size];
 
-    // Sort MethodDescs using the item kind for hot-cold spliting
-    struct SortMethodDesc : CQuickSort< MethodDesc * >
+    for (COUNT_T i = 0; i < count; i++)
     {
-        DataImage * m_image;
-
-        SortMethodDesc(DataImage *image, MethodDesc **pBase, SSIZE_T iCount)
-            : CQuickSort< MethodDesc * >(pBase, iCount),
-            m_image(image)
-        {
-        }
-
-        int Compare(MethodDesc ** ppMD1, MethodDesc ** ppMD2)
-        {
-            MethodDesc * pMD1 = *ppMD1;
-            MethodDesc * pMD2 = *ppMD2;
-
-            // Compare item kind
-            DataImage::ItemKind kind1 = GetPrecodeItemKind(m_image, pMD1);
-            DataImage::ItemKind kind2 = GetPrecodeItemKind(m_image, pMD2);
+        MethodDesc * pMD = rgMD[i];
+        StubPrecode * pPrecode = (StubPrecode *)(pBase + (sizeOfOne * i));
 
-            return kind1 - kind2;
-        }
-    };
+        pPrecode->Init(pMD, NULL);
+        _ASSERTE((Precode *)pPrecode == Precode::GetPrecodeForTemporaryEntryPoint(pBase, i));
 
-    SortMethodDesc sort(image, &(m_rgPendingChunk[0]), m_rgPendingChunk.GetCount());
-    sort.Sort();
+        // Alias the temporary entrypoint to the "new" MD - the one saved in the image.
+        image->RegisterSurrogate((MethodDesc *)image->GetImagePointer(pMD), pPrecode);
 
-    DataImage::ItemKind pendingKind = DataImage::ITEM_METHOD_PRECODE_COLD_WRITEABLE;
-    COUNT_T pendingCount = 0;
+        // If asked for, register alias the precode to the "old" MD (the one created from metadata) as well.
+        if (rgRegisterSurrogate[i])
+            image->RegisterSurrogate(pMD, pPrecode);
+    }
 
-    COUNT_T i;
-    for (i = 0; i < m_rgPendingChunk.GetCount(); i++)
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+    image->SaveStubPrecodeChunk(pBase, sizeOfOne, rgMD, count, kind);
+#else
+    ZapStoredStructure * pNode = image->StoreStructure(NULL, size, kind, PRECODE_ALIGNMENT);
+    for (COUNT_T i = 0; i < count; i++)
     {
-        MethodDesc * pMD = m_rgPendingChunk[i];
+        image->BindPointer((void *)(pBase + (sizeOfOne * i)), pNode, sizeOfOne * i);
+    }
 
-        DataImage::ItemKind kind = GetPrecodeItemKind(image, pMD);
-        if (kind != pendingKind)
-        {
-            if (pendingCount != 0)
-                SaveFixupPrecodeChunk(image, &(m_rgPendingChunk[i-pendingCount]), pendingCount, pendingKind);
+    image->CopyData(pNode, (void *)pBase, size);
+#endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
 
-            pendingKind = kind;
-            pendingCount = 0;
-        }
+    return (PVOID)pBase;
+}
 
-        pendingCount++;
-    }
+PVOID Precode::SaveChunk::Save(DataImage * image)
+{
```

Comment:
I think you will hit a problem with remoting on desktop here.

RemotingPrecode takes early out in the case the method is intercepted for remoting. The method itself won't even get JITed or restored from NGen image. I do not think that it will be the case with these changes because of there won't be remoting precode to do the interception. We will hit the prestub that will try to JIT/restore the method. JITing/restoring of the method is a problem because of it will try to load additional dependencies that may fail. There are real-world scenarios that will break because of this.

Reply:
Just so that I understand this - you're saying that for remoting we need to keep saving RemotingPrecodes even though they can't be patched (and thus if we did get to PreStub for them, we would have to create a new RemotingPrecode for the method anyway)... right?

Reply:
Just curious how this works for runtime loaded stuff... since we can't really have remoting precodes in temporary entry points (even at runtime temporary entry points are fixup precodes (or stub or compact))...
Does it mean that for these methods we create a stable entry point with the RemotingPrecode?

Reply:
>  you're saying that for remoting we need to keep saving RemotingPrecodes even though they can't be patched

Right. Or do surgery on prestub to just generate remoting precode for the remoting interception, without JITing the method.

> since we can't really have remoting precodes in temporary entry points

We do not hand out the temporary entrypoints for these. I think that the following piece in TryGetMultiCallableAddrOfCode takes care of it as side-effect:

```
    // Force the creation of the precode if we would eventually got one anyway
    if (MayHavePrecode())
        return GetOrCreatePrecode()->GetEntryPoint();
```

Reply:
I implemented the ability to save a specific precode (remoting for now) to a given method on top of the temporary entry point (so the method will have both, with the specific one used in MT slots).

Reply:
> so the method will have both, with the specific one used in MT slots

How is the slot patching going to work with this? The slot is going to point to the remoting precode. The patching expects the slot to point to the temporary entrypoint. So it won't ever get patched, I think.

Reply:
Thinking about this some more - maybe the surgery on the prestub would be easier way to deal with this. It should be pretty minor:  You really just need to call `GetOrCreatePrecode` and return if you get called with transparent proxy methodtable.

Reply:
Reverted the change for specific precodes and instead added a rather simple branch into DoPrestub.